plugins/migration/bin/rename.js

```diff
@@ -0,0 +1,208 @@
+   */
+  static calculateRenamings(database, currVersion, newVersion) {
+    currVersion = versionUtils.coerce(currVersion);
+    newVersion = versionUtils.coerce(newVersion);
+    const versions = Object.keys(database).sort(versionUtils.compare);
+    const renamers /** !Array<!VersionRenamer> */ = [];
+    for (const version of versions) {
+      // Only process versions in the range (currVersion, ^newVersion].
+      if (versionUtils.lte(version, currVersion)) continue;
+      if (versionUtils.gt(version, newVersion)) break;
+
+      renamers.push(new VersionRenamer(database[version]));
+    }
+    return renamers;
+  }
+
+  /**
+   * Applies the given renamings directly to a JavaScript string
+   * (presumably the contents of a developer's file).
+   * @param {string} str The string to apply the renamings in.
+   * @return {string} The string with renamings applied.
+   */
+  rename(str) {
+    return str.replace(dottedIdentifier, (match) => {
+      for (const versionRenamer of this.versionRenamers_) {
+        match = versionRenamer.rename(match);
+      }
+      return match;
+    });
+  }
+}
+
+/**
+ * An object which can apply a given set of renamings for a single
+ * Blockly version to the individual dotted-identifier paths.
+ */
+class VersionRenamer {
+  /**
+   * Constructor for the version renamer.
+   * @param {!Object} entry The database entry for a single version.
+   */
+  constructor(entry) {
+    /**
+     * List of pre-compiled renamings.
+     * @private @const {
+     *     !Array<{old: string, new: ?string, get: ?string, set: ?string}>}
+     */
+    this.renamings_ = [];
+    // See the sample entry in renamings.json5 for explanation of the
+    // meaning of the different properties on database entries.
+    for (const module of entry) {
+      // TODO: switch all of the || to ?? if/when we drop Node12.
+      const oldModulePath = module.oldPath || module.oldName;
+      const newExport = module.newExport ? '.' + module.newExport : '';
+      const newModulePath =
+          module.newPath || (module.newName || oldModulePath) + newExport;
+
+      if (module.exports) {
+        for (const [oldExportName, info] of Object.entries(module.exports)) {
+          const oldExportPath =
+              info.oldPath || `${oldModulePath}.${oldExportName}`;
+          const newBase = (info.newModule || newModulePath) + '.';
+          const renaming = {old: oldExportPath};
+          if (info.newPath) { // If newPath provided just use that.
+            renaming.new = info.newPath;
+          } else if (info.getMethod || info.setMethod) {
+            renaming.get = info.getMethod ? newBase + info.getMethod : null;
+            renaming.set = info.setMethod ? newBase + info.setMethod : null;
+          } else {
+            renaming.new = newBase + (info.newExport || oldExportName);
+          }
+          this.renamings_.push(renaming);
+        }
+      }
+
+      this.renamings_.push({old: oldModulePath, new: newModulePath});
```

Comment:
I expected this to be directly after the lines that construct `oldModulePath` and `newModulePath`. is it down here because the order of the renamings matters?

Also, why is it safe to use `new` as a key in your object here? It's normally a reserved keyword.

Reply:
I believe it is down here because we need to rename properties before renaming modules, otherwise we won't be able to properly detect properties that need to be renamed.

Eg:
moduleA -> moduleB
moduleA.exportA -> moduleC.exportC

If we rename moduleA first we won't be able to find moduleA.exportA because it will have become moduleB.exportA

@cpcallen  to check. If I'm right, I'll add a comment.

> Also, why is it safe to use new as a key in your object here? It's normally a reserved keyword.

I don't think there's any conflict because we're using it as a property, not a local identifier. Different grammar rules afaik.